air/src/stack/mod.rs

```diff
@@ -91,6 +116,206 @@ pub fn enforce_unique_constraints<E: FieldElement>(
     constraint_offset
 }
 
+/// Enforces generic constraints of all the stack ops.
+pub fn enforce_generic_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: &op_flags::OpFlags<E>,
+) -> usize {
+    let mut constraint_offset = 0;
+
+    // generic constraints of operations which doesnt shift the stack .
+    constraint_offset += no_shift_constraints(frame, result, op_flag);
+
+    // generic constraints of operations which shifts the stack to the left.
+    constraint_offset += left_shift_constraints(frame, &mut result[constraint_offset..], op_flag);
+
+    // generic constraints of operations which shifts the stack to the right.
+    constraint_offset += right_shift_constraints(frame, &mut result[constraint_offset..], op_flag);
+
+    constraint_offset
+}
+
+/// Enforces generic constraints of all the stack ops which doesn't shift the stack.
+fn no_shift_constraints<E: FieldElement>(
+    frame: &EvaluationFrame<E>,
+    result: &mut [E],
+    op_flag: &op_flags::OpFlags<E>,
+) -> usize {
+    // enforces the 1st item of the stack is copied over.
+    result[0] = op_flag.no_change_0() * are_equal(frame.stack_item(0), frame.stack_item_next(0));
+
+    // enforces the 2nd item of the stack is copied over.
+    result[1] = op_flag.no_change_1() * are_equal(frame.stack_item(1), frame.stack_item_next(1));
```

Comment:
There is one other way to do this which should reduce the number of constraints 3x. Specifically, we should be able to combine constraints for the same stack item into a single expression:

```
let expected_next_item_1 = op_flag.no_change_1() * frame.stack_item(1)
                         + op_flag.left_shift_2() * frame.stack_item(2)
                         + op_flag.right_shift_0() * frame.stack_item(0);
result[1] = are_equal(frame.stack_item_next(1), expected_next_item_1);
```

Reply:
Yes, indeed! Thanks, I missed this. I've added it in the commit [bc6237d](https://github.com/maticnetwork/miden/pull/392/commits/bc6237dc9f64b13234fc92e1ef964899d3ed2913).